### PR TITLE
Feature: Reveal chat history when chat message box is open.

### PR DESCRIPTION
**Pull request by PeterN**
## Motivation / Problem

Chat history is removed from the screen after a few seconds. It is visible by opening the console, but this fills a large portion of the screen and may also contain other non-chat messages. It also requires extra keypresses.

## Description

Instead of removing chat messages after the chat duration, they are now hidden from the normal chat history, until the chat message box is opened, where the full message history is automatically revealed.

## Limitations

Default chat history length seems to be quite long by default.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/network/network_chat_gui.cpp [hunk 1]
```diff
@@ -44,9 +45,10 @@ struct ChatMessage {
 };
 
 /* used for chat window */
-static ChatMessage *_chatmsg_list = nullptr; ///< The actual chat message list.
+static std::deque<ChatMessage> _chatmsg_list; ///< Chat message list.
 static bool _chatmessage_dirty = false;   ///< Does the chat message need repainting?
 static bool _chatmessage_visible = false; ///< Is a chat message visible.
+static std::chrono::steady_clock::time_point _chatmessage_dirty_time;
```
Comment (stormcone) [inline src/network/network_chat_gui.cpp]: Missing documentation.

File: src/network/network_chat_gui.cpp [hunk 2]
```diff
@@ -61,14 +63,17 @@ static uint8 *_chatmessage_backup = nullptr; ///< Backup in case text is moved.
  * Count the chat messages.
  * @return The number of chat messages.
  */
-static inline uint GetChatMessageCount()
+static inline uint GetChatMessageCount(bool show_all)
```
Comment (stormcone) [inline src/network/network_chat_gui.cpp]: Please update the documentation with the new argument.

**Review by stormcone — COMMENTED** [review]
Some minor thingie. :)

File: src/network/network_chat_gui.cpp [hunk 3]
```diff
@@ -44,12 +45,18 @@ struct ChatMessage {
...
 static bool _chatmessage_dirty = false;   ///< Does the chat message need repainting?
 static bool _chatmessage_visible = false; ///< Is a chat message visible.
 static bool _chat_tab_completion_active;  ///< Whether tab completion is active.
 static uint MAX_CHAT_MESSAGES = 0;        ///< The limit of chat messages to show.
 
+/**
+ * Time the chat history was marked dirty. This is used to determine if expired
+ * messages have recently expired and should cause a redraw to hide them.
+ */
+static std::chrono::steady_clock::time_point _chatmessage_dirty_time;
```
Comment (LordAro) [inline src/network/network_chat_gui.cpp:58]: Might be worth combining these all into a struct? Just to hold everything together

File: src/network/network_chat_gui.cpp [hunk 4]
```diff
@@ -229,9 +231,11 @@ void NetworkDrawChatMessage()
 
 	_cur_dpi = &_screen; // switch to _screen painting
 
+	auto now = std::chrono::steady_clock::now();
 	int string_height = 0;
-	for (uint i = 0; i < count; i++) {
-		SetDParamStr(0, _chatmsg_list[i].message);
+	for (auto iter = _chatmsg_list.begin(); iter != _chatmsg_list.end(); iter++) {
```
Comment (LordAro) [inline src/network/network_chat_gui.cpp]: ranged for?

File: src/network/network_chat_gui.cpp [hunk 5]
```diff
@@ -168,28 +174,21 @@ void NetworkUndrawChatMessage()
...
 
 /** Check if a message is expired. */
 void NetworkChatMessageLoop()
 {
-	for (uint i = 0; i < MAX_CHAT_MESSAGES; i++) {
-		ChatMessage *cmsg = &_chatmsg_list[i];
-		if (cmsg->message[0] == '\0') continue;
-
+	auto now = std::chrono::steady_clock::now();
+	for (auto it = _chatmsg_list.begin(); it != _chatmsg_list.end(); it++) {
```
Comment (LordAro) [inline src/network/network_chat_gui.cpp]: ranged for?

**Comment by PeterN** [comment]
Tweaked to keep list in the order it is displayed, so no reverse iteration is required.

File: src/network/network_chat_gui.cpp [hunk 6]
```diff
@@ -58,17 +65,20 @@ static PointDimension _chatmsg_box;
...
-		if (_chatmsg_list[i].message[0] == '\0') break;
+	if (show_all) return _chatmsg_list.size() != 0;
+
+	auto now = std::chrono::steady_clock::now();
+	for (auto cmsg : _chatmsg_list) {
+		if (cmsg.remove_time >= now) return true;
 	}
 
-	return i;
+	return false;
```
Comment (TrueBrain) [inline src/network/network_chat_gui.cpp:81]: `uint` returning `false` :)

File: src/network/network_chat_gui.cpp [hunk 7]
```diff
@@ -168,28 +173,21 @@ void NetworkUndrawChatMessage()
...
 
 /** Check if a message is expired. */
 void NetworkChatMessageLoop()
 {
-	for (uint i = 0; i < MAX_CHAT_MESSAGES; i++) {
-		ChatMessage *cmsg = &_chatmsg_list[i];
-		if (cmsg->message[0] == '\0') continue;
-
+	auto now = std::chrono::steady_clock::now();
+	for (auto it = _chatmsg_list.begin(); it != _chatmsg_list.end(); it++) {
```
Comment (TrueBrain) [inline src/network/network_chat_gui.cpp]: `for (auto it : _chatmsg_list)` ?

Learning C++, I can be dead-wrong.

Comment (PeterN) [reply src/network/network_chat_gui.cpp]: Lost in a rebase conflict, oops.

File: src/network/network_chat_gui.cpp [hunk 8]
```diff
@@ -58,17 +65,20 @@ static PointDimension _chatmsg_box;
 static uint8 *_chatmessage_backup = nullptr; ///< Backup in case text is moved.
 
 /**
- * Count the chat messages.
- * @return The number of chat messages.
+ * Test if there are any chat messages to display.
+ * @param show_all Set if all messages should be included, instead of unexpired only.
+ * @return True iff there are chat messages to display.
```
Comment (stormcone) [inline src/network/network_chat_gui.cpp:70]:
```suggestion
 * @return True if there are chat messages to display.
```
:)

Comment (TrueBrain) [reply src/network/network_chat_gui.cpp:70]: "Usage of the abbreviation "iff" first appeared in print in John L. Kelley's 1955 book General Topology. Its invention is often credited to Paul Halmos, who wrote "I invented 'iff,' for 'if and only if'—but I could never believe I was really its first inventor.""

:)

**Review by LordAro — APPROVED** [review]
Sure.

**Pull request merged**